mtags/src/main/scala-3/scala/meta/internal/pc/SignatureHelpProvider.scala

```diff
@@ -0,0 +1,163 @@
+package scala.meta.internal.pc
+
+import scala.collection.JavaConverters.*
+
+import scala.meta.internal.mtags.MtagsEnrichments.*
+import scala.meta.pc.OffsetParams
+
+import dotty.tools.dotc.ast.Trees.UnApply
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.ast.tpd.Tree
+import dotty.tools.dotc.core.Constants.Constant
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Denotations.SingleDenotation
+import dotty.tools.dotc.core.Symbols
+import dotty.tools.dotc.core.Types.ErrorType
+import dotty.tools.dotc.core.Types.MethodType
+import dotty.tools.dotc.interactive.Interactive
+import dotty.tools.dotc.interactive.InteractiveDriver
+import dotty.tools.dotc.reporting.AmbiguousOverload
+import dotty.tools.dotc.reporting.NoMatchingOverload
+import dotty.tools.dotc.util.Signatures
+import dotty.tools.dotc.util.SourcePosition
+import dotty.tools.dotc.util.Spans.Span
+import org.eclipse.lsp4j.ParameterInformation
+import org.eclipse.lsp4j.SignatureHelp
+import org.eclipse.lsp4j.SignatureInformation
+
+object SignatureHelpProvider:
+
+  def contribute(driver: InteractiveDriver, params: OffsetParams) =
+    val uri = params.uri
+    val sourceFile = CompilerInterfaces.toSource(params.uri, params.text)
+    driver.run(uri, sourceFile)
+
+    given ctx: Context = driver.currentCtx
+    // given locatedCtx: Context = driver.localContext(params)
+    val pos = driver.sourcePosition(params)
+    val trees = driver.openedTrees(uri)
+    val path =
+      Interactive
+        .pathTo(trees, pos)(using ctx)
+        .dropWhile {
+          case _: tpd.Apply => false
+          case unapply: tpd.UnApply =>
+            // Don't show tuple unapply method
+            val sym = unapply.fun.symbol.owner.companionClass
+            Symbols.defn.isTupleClass(sym)
+          case _ => true
+        }
+
+    val (paramN, callableN, alternatives) =
+      callInfo(path, pos.span)(using ctx)
+
+    val signatureInfos =
+      alternatives.flatMap(Signatures.toSignature(_)(using ctx))
+    new SignatureHelp(
+      signatureInfos.map(signatureToSignatureInformation).asJava,
+      callableN,
+      paramN
+    )
+  end contribute
+
+  private def callInfo(path: List[Tree], span: Span)(using
+      ctx: Context
+  ): (Int, Int, List[SingleDenotation]) =
+    path match
+      case (tpd.UnApply(fun, implicits, patterns)) :: _ =>
```

Comment:
This is the main fix, we apply almost the same logic as for Apply in `Signatures`

Reply:
Are we going to remove this pattern match once https://github.com/lampepfl/dotty/pull/14611 merged? In this case, can we wait for the next release that contains https://github.com/lampepfl/dotty/pull/14611 ?


Or leave this method for older Scala3 compilers? In this case, can we add a comment on why there's a kind of duplicated logic in metals and dotty?

Reply:
I think we should leave it as a fallback, but I still need to work a bit on this.